### PR TITLE
Modules

### PR DESCRIPTION
**Pull request by tomchop**
This PR removes collectors, processors, and exporters in favor a more generic `BaseModule`. This removes quite a lot of duplicated code and increases flexibility.

It also introduces a `DFTimewolfState` object which enables better error management and handles forwarding of data between modules.

### CODE REVIEW

File: dftimewolf/lib/state.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+"""This class maintains the internal dfTimewolf state.
```
Comment (berggren) [inline dftimewolf/lib/state.py]: Seeing that this new state machine introduces complexity, I opt for some design/documentation.

File: dftimewolf/lib/state.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    """
+    # Move any existing errors to global errors
+    self.global_errors.extend(self.errors)
+    self.errors = []
+
+    # Make the previous module's output available to the next module
+    self.input = self.output
+    self.output = []
+
+  def check_errors(self, global_=False):
```
Comment (berggren) [inline dftimewolf/lib/state.py]: I would skip the trailing _

Comment (berggren) [reply dftimewolf/lib/state.py]: maybe global_error instead?

File: dftimewolf/lib/state.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+Use it to track errors, abort on global failures, cleanup after modules, etc.
+"""
+
+import sys
+
+class DFTimewolfState(object):
+  """The main State class.
+
+  Attributes:
+    errors: [(str, bool)] The errors generated by a module. These should be
```
Comment (berggren) [inline dftimewolf/lib/state.py]: The type hinting in the docstrings are inconsistent with the rest of the codebase. Maybe put in a TODO to refactor the rest as well (in a later PR)

File: dftimewolf/cli/dftimewolf_recipes.py [hunk 4]
```diff
@@ -94,98 +75,28 @@ def main():
...
-    console_out.StdOut(
-        'Recipe {0:s} executed with {1:d} errors:'.format(
-            recipe['name'], len(global_errors)))
-    for error in global_errors:
-      console_out.StdOut('  {0:s}: {1:s}'.format(*error))
+        module_description['args'], vars(args), config.Config)
+
+    # Create the module object and start processing
+    module_name = module_description['name']
+    print 'Running module {0:s}'.format(module_name)
```
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/cli/dftimewolf_recipes.py]: I'd like to start with prints and implement better logging in a future PR. The logging function that existed was a bit too clumsy to use IMO.

File: dftimewolf/config.py [hunk 5]
```diff
@@ -125,46 +123,4 @@ def get_collector(cls, name):
     Returns:
       Corresponding class extending BaseCollector.
```
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/config.py]: I'm sorry, this should be `BaseModule`. Collectors don't exist anymore

File: dftimewolf/lib/collectors/filesystem.py [hunk 6]
```diff
@@ -5,77 +5,39 @@
 
 import os
 
-from dftimewolf.lib.collectors.collectors import BaseCollector
+from dftimewolf.lib.module import BaseModule
```
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/collectors/filesystem.py]: Same, we're not talking about collectors, processors, or exporters anymore but generic modules.

File: dftimewolf/lib/state.py (hunk 1 above)
Comment (tomchop) [reply dftimewolf/lib/state.py]: Sounds good. Where do we put it?

File: dftimewolf/lib/state.py (hunk 2 above)
Comment (tomchop) [reply dftimewolf/lib/state.py]: I'll change it to `is_global`

File: dftimewolf/lib/state.py (hunk 3 above)
Comment (tomchop) [reply dftimewolf/lib/state.py]: Will do.

File: dftimewolf/lib/collectors/filesystem.py [hunk 7]
```diff
@@ -5,77 +5,39 @@
...
+    """Sets up the _paths attribute
 
     Args:
-      paths: List of strings representing paths to collect files from
-      verbose: Print extra output to stdout (default: False)
-
-    Returns:
-      A list of FilesystemCollector objects that can be join()ed from the
-      caller.
+      paths: Comma-separated list of strings represnting the paths to collect.
```
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/collectors/filesystem.py]: Ack.

File: dftimewolf/lib/state.py (hunk 1 above)
Comment (berggren) [reply dftimewolf/lib/state.py]: I suggest the github wiki, under developer docs.

File: dftimewolf/lib/exporters/timesketch.py [hunk 8]
```diff
@@ -1,104 +1,84 @@
...
+      sketch_description = 'Sketch generated by dfTimewolf'
+
       self.sketch_id = self.timesketch_api.create_sketch(
           sketch_name, sketch_description)
-      self.console_out.StdOut(
-          'New sketch created: {0:d}'.format(self.sketch_id))
-      syslog.syslog('New sketch created: {0:d}'.format(self.sketch_id))
+      message = 'Sketch {0:d} created'.format(self.sketch_id)
+      print message
+      syslog.syslog(message)
```
[the remark this answers is not in the document]
Comment (tomchop) [reply dftimewolf/lib/exporters/timesketch.py]: This is a remnant of the old logging / stdout system. I'll remove it and keep the print for now. I'll add an issue and TODOs to get the logging fixed.

**Pull request merged**